File: practice/educational_DP/2nd.cpp
#include <bits/stdc++.h>
using ll = long long int;
using namespace std;

int main()
{
    int N, K;
    cin >> N >> K;
    vector<int> H(N, 0);
    vector<ll> dp(N + 1, 1 << 30);
    for (int i = 0; i < N; i++)
    {
        cin >> H[i];
    }
    dp[0] = 0;
    for (int i = 0; i < N; i++)
    {
        for (int j = 1; j <= K; j++)
        {
            if (i + j <= N)
            {
                dp[i + j] = min(dp[i + j], dp[i] + abs(H[i] - H[i + j]));
            }
        }
    }
    cout << dp[N - 1] << endl;

    return 0;
}

File: practice/educational_DP/2nd.py
N, K = map(int, input().split())
H = list(map(int, input().split()))
dp = [10**10 for i in range(0, N+1)]
dp[0] = 0
for i in range(1, N):
    for j in range(1, K+1):
        if i-j >= 0:
            dp[i] = min(dp[i], dp[i-j]+abs(H[i]-H[i-j]))
print(dp[N-1])
